File: GXX.CSharp/src/GXX.RunGate/BagItemList.cs
using System;
using System.Collections.Generic;
using System.Threading;

namespace GXX.RunGate;

/// <summary>TBagItem（BagItemList.pas packed 记录）。</summary>
public class TBagItem
{
    public int MakeIndex;
    public byte StdMode;
    public byte Shape;
    public ushort Reserved;
    public int AC1;
    public int MAC1;
}

/// <summary>
/// BagItemList.pas TBagItemList 1:1：按 MakeIndex 有序的背包物品索引表
/// （二分查找 + 插入排序语义 + 临界锁 + 可选重复项）。
/// </summary>
public class TBagItemList
{
    private readonly List<TBagItem> _list = new();
    private readonly object _locker = new();
    private bool _duplicates;

    public TBagItemList(int capacity)
    {
        _list.Capacity = Math.Max(capacity, 0);
    }

    public int Count
    {
        get { lock (_locker) return _list.Count; }
    }

    public int Capacity
    {
        get { lock (_locker) return _list.Capacity; }
        set { lock (_locker) _list.Capacity = value; }
    }

    public bool Duplicates { get => _duplicates; set => _duplicates = value; }

    public void Lock() => System.Threading.Monitor.Enter(_locker);
    public void UnLock() => System.Threading.Monitor.Exit(_locker);

    /// <summary>二分查找：返回是否存在与插入位置（原 Search 语义）。</summary>
    private bool Search(int makeIndex, out int index)
    {
        int lo = 0, hi = _list.Count - 1;
        while (lo <= hi)
        {
            int mid = (lo + hi) / 2;
            int midIdx = _list[mid].MakeIndex;
            if (makeIndex == midIdx) { index = mid; return true; }
            if (makeIndex < midIdx) hi = mid - 1;
            else lo = mid + 1;
        }
        index = lo;
        return false;
    }

    /// <summary>Add：已存在（且不允许重复）→ 返回原项；否则插入有序位置。</summary>
    public TBagItem Add(int makeIndex)
    {
        lock (_locker)
        {
            if (!_duplicates && Search(makeIndex, out int idx))
                return _list[idx];
            var item = new TBagItem { MakeIndex = makeIndex };
            Search(makeIndex, out int insertPos);
            _list.Insert(_duplicates ? Count : insertPos, item);
            return item;
        }
    }

    public TBagItem? Find(int makeIndex)
    {
        lock (_locker)
        {
            if (Search(makeIndex, out int idx))
                return _list[idx];
            return null;
        }
    }

    public void Clear()
    {
        lock (_locker) _list.Clear();
    }

    public void Delete(int index)
    {
        lock (_locker) _list.RemoveAt(index);
    }

    public void Remove(int makeIndex)
    {
        lock (_locker)
        {
            if (Search(makeIndex, out int idx))
                _list.RemoveAt(idx);
        }
    }

    public void Remove(TBagItem bagItem)
    {
        lock (_locker) _list.Remove(bagItem);
    }
}

File: Source/RunGate/BagItemList.pas
unit BagItemList;

interface

{$I Iocp.inc}

uses
  Classes, SysUtils, IocpCommon;

type
  PBagItem = ^TBagItem;
  TBagItem = packed record
    MakeIndex: Integer;
    StdMode: Byte;
    Shape: Byte;
    Reserved: Word;
    //Name: string[30];
    AC1: Integer;
    MAC1: Integer;
  end;

  TBagItemList = class(TObject)
  private
    FDuplicates: Boolean;
    FList: TList;
    FLocker: TIocpCriticalSection;
    function GetCapacity: Integer;
    //function GetCount: Integer;
    //function GetItems(Index: Integer): PBagItem;
    procedure SetCapacity(const Value: Integer);
  protected
    function Search(MakeIndex: Integer; var Index: Integer): Boolean; virtual;
  public
    constructor Create(ACapacity: Integer); virtual;
    destructor Destroy; override;

    function Add(MakeIndex: Integer): PBagItem; virtual;
    function Find(MakeIndex: Integer): PBagItem;

    procedure Clear; virtual;
    procedure Delete(Index: Integer);
    procedure Remove(MakeIndex: Integer); overload;
    procedure Remove(BagItem: PBagItem); overload;

    property Capacity: Integer read GetCapacity write SetCapacity;
    //property Count: Integer read GetCount;
    //property Items[Index: Integer]: PBagItem read GetItems; default;
    property Duplicates: Boolean read FDuplicates write FDuplicates;

    procedure Lock;
    procedure UnLock;
  end;

implementation

{ TBagItemList }

constructor TBagItemList.Create(ACapacity: Integer);
begin
  FLocker := TIocpCriticalSection.Create({$IFDEF USE_SPINLOCK}'TBagItemList.FLocker'{$ENDIF});
  FList := TList.Create;
  FList.Capacity := ACapacity;
  FDuplicates := False;
end;

function TBagItemList.Add(MakeIndex: Integer): PBagItem;
var
  Index: Integer;
begin
  if Search(MakeIndex, Index) then
  begin
    Result := FList.Items[Index];
  end
  else
  begin
    New(Result);
    Result.MakeIndex := MakeIndex;
    FList.Insert(Index, Result);
  end;
end;

procedure TBagItemList.Clear;
var
  I: Integer;
  BagItem: PBagItem;
begin
  for I := 0 to FList.Count - 1 do
  begin
    BagItem := FList.Items[I];
    Dispose(BagItem);
  end;
  FList.Clear;
end;

procedure TBagItemList.Delete(Index: Integer);
begin
  if (Index >= 0) and (Index < FList.Count) then
  begin
    Dispose(PBagItem(FList.Items[Index]));
    FList.Delete(Index);
  end;
end;

destructor TBagItemList.Destroy;
begin
  Clear;
  FList.Free;
  FLocker.Free;
  inherited;
end;

function TBagItemList.GetCapacity: Integer;
begin
  Result := FList.Capacity;
end;

{
function TBagItemList.GetCount: Integer;
begin
  Result := FList.Count;
end;

function TBagItemList.GetItems(Index: Integer): PBagItem;
begin
  if (Index >= 0) and (Index < FList.Count) then
    Result := FList.Items[Index]
  else
    Result := nil;
end;
}

procedure TBagItemList.Remove(MakeIndex: Integer);
var
  Index: Integer;
begin
  if Search(MakeIndex, Index) then
  begin
    Dispose(PBagItem(FList.Items[Index]));
    FList.Delete(Index);
  end;
end;

procedure TBagItemList.Remove(BagItem: PBagItem);
begin
  FList.Remove(BagItem);
  Dispose(BagItem);
end;

function TBagItemList.Search(MakeIndex: Integer;
  var Index: Integer): Boolean;
var
  L, H, I, C: Integer;
begin
  Result := False;
  L := 0;
  H := FList.Count - 1;
  while L <= H do
  begin
    I := L + (H - L) shr 2;
    C := PBagItem(FList.Items[I]).MakeIndex - MakeIndex;
    if C < 0 then
      L := I + 1
    else
    begin
      H := I - 1;
      if C = 0 then
      begin
        Result := True;
        if not Duplicates then L := I;
      end;
    end;
  end;
  Index := L;
end;

procedure TBagItemList.SetCapacity(const Value: Integer);
begin
  FList.Capacity := Value;
end;

function TBagItemList.Find(MakeIndex: Integer): PBagItem;
var
  Index: Integer;
begin
  if Search(MakeIndex, Index) then
    Result := FList.Items[Index]
  else
    Result := nil;
end;

procedure TBagItemList.Lock;
begin
  FLocker.Lock;
end;

procedure TBagItemList.UnLock;
begin
  FLocker.UnLock;
end;

end.
